src/NuGet.Core/NuGet.ProjectModel/LockFile/Utf8JsonStreamLockFileTargetLibraryConverter.cs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Text;
+using System.Text.Json;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// A <see cref="Utf8JsonStreamReaderConverter{T}"/> to allow read JSON into <see cref="LockFileTargetLibrary"/>
+    /// </summary>
+    internal class Utf8JsonStreamLockFileTargetLibraryConverter : Utf8JsonStreamReaderConverter<LockFileTargetLibrary>
+    {
+        private static readonly byte[] Utf8Type = Encoding.UTF8.GetBytes("type");
+        private static readonly byte[] Utf8Framework = Encoding.UTF8.GetBytes("framework");
+        private static readonly byte[] Utf8Dependencies = Encoding.UTF8.GetBytes("dependencies");
+        private static readonly byte[] Utf8FrameworkAssemblies = Encoding.UTF8.GetBytes("frameworkAssemblies");
+        private static readonly byte[] Utf8Runtime = Encoding.UTF8.GetBytes("runtime");
+        private static readonly byte[] Utf8Compile = Encoding.UTF8.GetBytes("compile");
+        private static readonly byte[] Utf8Resource = Encoding.UTF8.GetBytes("resource");
+        private static readonly byte[] Utf8Native = Encoding.UTF8.GetBytes("native");
+        private static readonly byte[] Utf8Build = Encoding.UTF8.GetBytes("build");
+        private static readonly byte[] Utf8BuildMultiTargeting = Encoding.UTF8.GetBytes("buildMultiTargeting");
+        private static readonly byte[] Utf8ContentFiles = Encoding.UTF8.GetBytes("contentFiles");
+        private static readonly byte[] Utf8RuntimeTargets = Encoding.UTF8.GetBytes("runtimeTargets");
+        private static readonly byte[] Utf8Tools = Encoding.UTF8.GetBytes("tools");
+        private static readonly byte[] Utf8Embed = Encoding.UTF8.GetBytes("embed");
+        private static readonly byte[] Utf8FrameworkReferences = Encoding.UTF8.GetBytes("frameworkReferences");
+
+        private IList<PackageDependency> ReadPackageDependencyList(ref Utf8JsonStreamReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                return new List<PackageDependency>(0);
+            }
+
+            var packageDependencies = new List<PackageDependency>();
+            while (reader.Read() && reader.TokenType == JsonTokenType.PropertyName)
+            {
+                string propertyName = reader.GetString();
+                string versionString = reader.ReadNextTokenAsString();
```

Comment:
I imagine version strings and ranges are the same throughout the assets file, can we dedupe them (via `Dictionary<UTF8String, VersionRange>` or something) so we're only parsing the same string once per file?

Reply:
I just looked at an assets file from https://github.com/dotnet/project-system/tree/main/src/Microsoft.VisualStudio.ProjectSystem.Managed, and `6.0.0` shows up 162 times.

Reply:
@davkean I'm trying to get better acquainted with Memory<T>. We have three scenarios where we can cache the values here. 
VersionRange, NuGetVersion, and NuGetFramework. 

For NuGetVersion, it looks like we're parsing it once we've already retrieved the string for the current token. So I was planning on using Dictionary<string, NuGetVersion>.

For NuGetFramework and VersionRange we could use the current ValueSpan/ValueSequence of the current token. So if we can leverage the readonlyspan<byte> we could do this check and get the corresponding object without allocating a string each time. I'm guessing that was the idea behind the suggestion of using Dictionary<Memory<T>, VersionRange>? Would that mean using Dictionary<Memory<char[]>, VersionRange>, then using an IEqualityComparer to compare the arrays are equal with SequenceEqual?

I found this [open issue](https://github.com/dotnet/runtime/issues/2010) that looks like it's trying to address this.

Reply:
NuGetFramework already has some optimizations, https://github.com/NuGet/NuGet.Client/blob/fefd3eb3b77936a685260e26dad510b56ac7e66e/src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs#L533. This caches well known frameworks. 

Nothing for version range/version though.

Reply:
Okay, so framework might not be worth it unless we want to figure out how to avoid the extra string allocation.